packages/devtools_app/lib/src/screens/performance/panes/timeline_events/perfetto/_perfetto_controller_web.dart

```diff
@@ -249,36 +295,39 @@ class PerfettoController extends DisposableController
     }
   }
 
-  Future<void> _perfettoIFrameReady() async {
-    if (_perfettoIFrame.contentWindow == null) {
-      await _perfettoIFrame.onLoad.first;
-      assert(
-        _perfettoIFrame.contentWindow != null,
-        'Something went wrong. The iFrame\'s contentWindow is null after the'
-        ' onLoad event.',
-      );
-    }
-  }
-
   Future<void> _pingPerfettoUntilReady() async {
-    while (!_perfettoReady.isCompleted) {
-      await Future.delayed(const Duration(microseconds: 100), () async {
+    if (!_perfettoHandlerReady.isCompleted) {
+      _pollForPerfettoHandlerReady =
+          Timer.periodic(const Duration(milliseconds: 200), (_) async {
         // Once the Perfetto UI is ready, Perfetto will receive this 'PING'
         // message and return a 'PONG' message, handled in [_handleMessage].
         _postMessage(_perfettoPing);
       });
+
+      // Timeout after [_pollUntilReadyTimeout] has passed.
+      await Future.any([
+        _perfettoHandlerReady.future,
+        Future.delayed(_pollUntilReadyTimeout),
+      ]).then((_) => _pollForPerfettoHandlerReady?.cancel());
```

Comment:
There's [`Future.timeout` method](https://api.dart.dev/be/176281/dart-async/Future/timeout.html) to that might make this timeout pattern a little bit simpler:

```dart
try {
  await _perfettoHandlerReady.future.timeout(_pollUntilReadyTimeout);
// maybe catch the TimeoutException or define a different `onTimeout` behavior above
// so. you don't need a try-catch at all?
} finally {
  _pollForPerfettoHandlerReady?.cancel();
}
```

Reply:
done